A80-pid.md

```diff
@@ -0,0 +1,290 @@
+
+The `callbacks` object, which is to be provided by the balancer builder, will implement the following interface (expressed in pseudo-code):
+
+```
+wrrCallbacks interface {
+  onSubchannelAdded(subchannelID int, data callbackData, conf callbackConfig)
+  onSubchannelRemoved(subchannelID int, data callbackData, conf callbackConfig)
+
+  // onLoadReport is called when a new load report is received for a given subchannel. 
+  // This function returns the new weight for a subchannel. If the returned value is -1,
+  // the subchannel should keep using the old value.
+  // onLoadReport won't be called during the blackout period.
+  onLoadReport(subchannelId int, report loadReport, data callbackData, conf callbackConfig) float
+
+  // onEDFSchedulerUpdate is called after the wrr balancer recreates the EDF scheduler.
+  onEDFSchedulerUpdate(data callbackData, conf callbackConfig)
+}
+```
+
+Here is how `pid` balancer implements `wrrCallbacks` interface.
+```
+func onSubchannelAdded(subchannelID int, data callbackData, conf callbackConfig) {
+  // Do nothing
+}
+
+func onSubchannelRemoved(subchannelID int, data callbackData, conf callbackConfig) {
+  // Remove subchannelID from two maps that store the value of last utilization
+  // and last applied weight per subchannel
+  delete(data.utilizationPerSubchannel, subchannelID)
+  delete(data.lastAppliedWeightPerSubchannel, subchannelID)
+}
+
+
+func onLoadReport(subchannelId int, load loadReport, data callbackData, conf callbackConfig) float {
+  utilization = load.ApplicationUtilization
+  if utilization == 0 {
+    utilization = load.CpuUtilization
+  }
+  if utilization == 0 || load.RpsFractional == 0 {
+    // Ignore empty load
+    return -1
+  }
+  errorRate = load.Eps / load.RpsFractional
+  useErrPenalty = errorRate > conf.ErrorUtilizationThreshold
+  if useErrPenalty {
+    utilization += errorRate * conf.ErrorUtilizationPenalty
+  }
+
+  // Ensure at least WeightUpdatePeriod has passed since the last update.
+  // Prevents corruption of PID controller's internal state, which could happen in the following cases:
+  // * If 2 updates are very close to each other in time, samplingInterval ~= 0 and signal ~= infinity.
+  // * If multiple updates happened during a single WeightUpdatePeriod, the actual weights are not applied,
+  // but the PID controller keeps growing the weights and it may easily pass the balancing point.
+  if time.Since(lastApplied) < conf.WeightUpdatePeriod {
+    return -1
+  }
+
+  // use value calculated in the onEDFSchedulerUpdate method
+  meanUtilization = data.meanUtilization
+
+  // call the PID controller to get the value of the control signal.
+  controlSignal = data.pidController.update({
```

Comment:
I like this idea, but I think it also depends on the decision we make about my proposal in this thread https://github.com/grpc/proposal/pull/430/files#r1761294507 If we update load reports to give us monotonically increasing CPU counter this should solve the problem with low QPS clients. 

Also I am fine with disabling/discouraging OOB. That's what we do as well in our infrastructure. Since we pair this solution with heavy subsetting it also helps us to ensure every client has fresh data about server utilization, even if the client has low QPS.